File: 213 House Robber II.js
/*
House Robber II
https://leetcode.com/problems/house-robber-ii/
*/
/**
 * @param {number[]} nums
 * @return {number}
 */
var rob = function(nums) {
    var arrayLength = nums.length;
    var i = 0;
    
    if(arrayLength === 0) return 0;
    if(arrayLength === 1) return nums[0];
    if(arrayLength === 2) return Math.max(nums[0],nums[1]);
    if(arrayLength === 3) return Math.max(nums[0],nums[1],nums[2]);
    
    // rob 1st, not 2nd, not last
    var three_steps = nums[0]; // max at i - 3
    var two_steps = Math.max(nums[0],nums[1]); // max at i - 2
    var one_step = Math.max(nums[0]+nums[2],nums[1]); // max at i - 1
    var maxIncome = one_step;
    
    for (i = 3; i < arrayLength - 1; i++) {
        maxIncome = Math.max(maxIncome, Math.max(three_steps,two_steps) + nums[i]);
        three_steps = two_steps;
        two_steps = one_step;
        one_step = maxIncome;
    }
    
    // not rob 1st, rob 2nd, rob last
    three_steps = 0;
    two_steps = nums[1];
    one_step = Math.max(nums[2],nums[1]);
    var maxIncome2 = one_step;
    
    for (i = 3; i < arrayLength; i++) {
        maxIncome2 = Math.max(maxIncome2, Math.max(three_steps,two_steps) + nums[i]);
        three_steps = two_steps;
        two_steps = one_step;
        one_step = maxIncome2;
    }
    
    return Math.max(maxIncome,maxIncome2);
};

var r = 0
var arr = new Array(1,2,3,4,5,1,2,3,4,5)
r = rob(arr)
console.log(r);

File: 213 House Robber II.py
'''
213. House Robber II
https://leetcode.com/problems/house-robber-ii/

You are a professional robber planning to rob houses along a street. Each house has a certain amount of money stashed.
All houses at this place are arranged in a circle. That means the first house is the neighbor of the last one.
Meanwhile, adjacent houses have security system connected and it will automatically contact the police
if two adjacent houses were broken into on the same night.

Given a list of non-negative integers representing the amount of money of each house,
determine the maximum amount of money you can rob tonight without alerting the police.

Example 1:

Input: [2,3,2]
Output: 3
Explanation: You cannot rob house 1 (money = 2) and then rob house 3 (money = 2),
             because they are adjacent houses.
Example 2:

Input: [1,2,3,1]
Output: 4
Explanation: Rob house 1 (money = 1) and then rob house 3 (money = 3).
             Total amount you can rob = 1 + 3 = 4.
'''
class Solution:
    def getMaxCash(self, nums: List[int], dp: List[int]) -> int:
        for i in range(2, len(nums)):
            di = i % 3
            dp[di] = max(dp[di-1], dp[di-2] + nums[i])
        return dp[(len(nums) - 1) % 3]
        
    def rob(self, nums: List[int]) -> int:
        if not nums:
            return 0
        elif len(nums) == 1:
            return nums[0]
 
        # Rob 1st house, not 2nd house, not last house
        dp0 = [nums[0], nums[0], None]
        # Rob 2nd house, not 1st house, potentially last house
        dp1 = [0, nums[1], None]

        return max(self.getMaxCash(nums[:-1], dp0), self.getMaxCash(nums, dp1))
